File: lua/test/action_entity_test.lua
-- Action entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("trello_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ActionEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Action(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["action"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:Action(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:Action(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = action_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "action." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_ACTION_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local action_ref01_ent = client:Action(nil)
    local action_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.action"), "action_ref01"))
    action_ref01_data["card_id"] = setup.idmap["card01"]
    action_ref01_data["id_action"] = setup.idmap["id_action01"]
    action_ref01_data["member_id"] = setup.idmap["member01"]
    action_ref01_data["organization_id"] = setup.idmap["organization01"]

    local action_ref01_data_result, err = action_ref01_ent:create(action_ref01_data, nil)
    assert.is_nil(err)
    action_ref01_data = helpers.to_map(type(action_ref01_data_result) == 'table' and action_ref01_data_result.data_get and action_ref01_data_result:data_get() or action_ref01_data_result)
    assert.is_not_nil(action_ref01_data)
    assert.is_not_nil(action_ref01_data["id"])

    -- LIST
    local action_ref01_match = {
      ["organization_id"] = setup.idmap["organization01"],
    }

    local action_ref01_list_result, err = action_ref01_ent:list(action_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(action_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(action_ref01_list_result),
      { id = action_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local action_ref01_data_up0_up = {
      id = action_ref01_data["id"],
    }

    local action_ref01_markdef_up0_name = "date"
    local action_ref01_markdef_up0_value = "Mark01-action_ref01_" .. tostring(setup.now)
    action_ref01_data_up0_up[action_ref01_markdef_up0_name] = action_ref01_markdef_up0_value

    local action_ref01_resdata_up0_result, err = action_ref01_ent:update(action_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local action_ref01_resdata_up0 = helpers.to_map(type(action_ref01_resdata_up0_result) == 'table' and action_ref01_resdata_up0_result.data_get and action_ref01_resdata_up0_result:data_get() or action_ref01_resdata_up0_result)
    assert.is_not_nil(action_ref01_resdata_up0)
    assert.are.equal(action_ref01_resdata_up0["id"], action_ref01_data_up0_up["id"])
    assert.are.equal(action_ref01_resdata_up0[action_ref01_markdef_up0_name], action_ref01_markdef_up0_value)

    -- LOAD
    local action_ref01_match_dt0 = {
      id = action_ref01_data["id"],
    }
    local action_ref01_data_dt0_loaded, err = action_ref01_ent:load(action_ref01_match_dt0, nil)
    assert.is_nil(err)
    local action_ref01_data_dt0_load_result = helpers.to_map(type(action_ref01_data_dt0_loaded) == 'table' and action_ref01_data_dt0_loaded.data_get and action_ref01_data_dt0_loaded:data_get() or action_ref01_data_dt0_loaded)
    assert.is_not_nil(action_ref01_data_dt0_load_result)
    assert.are.equal(action_ref01_data_dt0_load_result["id"], action_ref01_data["id"])

    -- REMOVE
    local action_ref01_match_rm0 = {
      id = action_ref01_data["id"],
    }
    local _, err = action_ref01_ent:remove(action_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local action_ref01_match_rt0 = {
      ["organization_id"] = setup.idmap["organization01"],
    }

    local action_ref01_list_rt0_result, err = action_ref01_ent:list(action_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(action_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(action_ref01_list_rt0_result),
      { id = action_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function action_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/action/ActionTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read action test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "action01", "action02", "action03", "board01", "board02", "board03", "card01", "card02", "card03", "list01", "list02", "list03", "member01", "member02", "member03", "organization01", "organization02", "organization03", "id_action01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("TRELLO_TEST_ACTION_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["TRELLO_TEST_ACTION_ENTID"] = idmap,
    ["TRELLO_TEST_LIVE"] = "FALSE",
    ["TRELLO_TEST_EXPLAIN"] = "FALSE",
    ["TRELLO_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["TRELLO_TEST_ACTION_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["TRELLO_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["TRELLO_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["TRELLO_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["TRELLO_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/action_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/trello-sdk/go"
	"github.com/voxgig-sdk/trello-sdk/go/core"

	vs "github.com/voxgig-sdk/trello-sdk/go/utility/struct"
)

func TestActionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Action(nil)
		if ent == nil {
			t.Fatal("expected non-nil ActionEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"action": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Action(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Action(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := actionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "action." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_ACTION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		actionRef01Ent := client.Action(nil)
		actionRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "action"}, setup.data), "action_ref01"))
		actionRef01Data["card_id"] = setup.idmap["card01"]
		actionRef01Data["id_action"] = setup.idmap["id_action01"]
		actionRef01Data["member_id"] = setup.idmap["member01"]
		actionRef01Data["organization_id"] = setup.idmap["organization01"]

		actionRef01DataResult, err := actionRef01Ent.Create(actionRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		actionRef01Data = core.ToMapAny(entityData(actionRef01DataResult))
		if actionRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if actionRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		actionRef01Match := map[string]any{
			"organization_id": setup.idmap["organization01"],
		}

		actionRef01ListResult, err := actionRef01Ent.List(actionRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		actionRef01List, actionRef01ListOk := actionRef01ListResult.([]any)
		if !actionRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", actionRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(actionRef01List), map[string]any{"id": actionRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		actionRef01DataUp0Up := map[string]any{
			"id": actionRef01Data["id"],
		}

		actionRef01MarkdefUp0Name := "date"
		actionRef01MarkdefUp0Value := fmt.Sprintf("Mark01-action_ref01_%d", setup.now)
		actionRef01DataUp0Up[actionRef01MarkdefUp0Name] = actionRef01MarkdefUp0Value

		actionRef01ResdataUp0Result, err := actionRef01Ent.Update(actionRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		actionRef01ResdataUp0 := core.ToMapAny(entityData(actionRef01ResdataUp0Result))
		if actionRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if actionRef01ResdataUp0["id"] != actionRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if actionRef01ResdataUp0[actionRef01MarkdefUp0Name] != actionRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", actionRef01MarkdefUp0Name, actionRef01ResdataUp0[actionRef01MarkdefUp0Name])
		}

		// LOAD
		actionRef01MatchDt0 := map[string]any{
			"id": actionRef01Data["id"],
		}
		actionRef01DataDt0Loaded, err := actionRef01Ent.Load(actionRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		actionRef01DataDt0LoadResult := core.ToMapAny(entityData(actionRef01DataDt0Loaded))
		if actionRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if actionRef01DataDt0LoadResult["id"] != actionRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		actionRef01MatchRm0 := map[string]any{
			"id": actionRef01Data["id"],
		}
		_, err = actionRef01Ent.Remove(actionRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		actionRef01MatchRt0 := map[string]any{
			"organization_id": setup.idmap["organization01"],
		}

		actionRef01ListRt0Result, err := actionRef01Ent.List(actionRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		actionRef01ListRt0, actionRef01ListRt0Ok := actionRef01ListRt0Result.([]any)
		if !actionRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", actionRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(actionRef01ListRt0), map[string]any{"id": actionRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func actionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "action", "ActionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read action test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse action test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"action01", "action02", "action03", "board01", "board02", "board03", "card01", "card02", "card03", "list01", "list02", "list03", "member01", "member02", "member03", "organization01", "organization02", "organization03", "id_action01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRELLO_TEST_ACTION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRELLO_TEST_ACTION_ENTID": idmap,
		"TRELLO_TEST_LIVE":      "FALSE",
		"TRELLO_TEST_EXPLAIN":   "FALSE",
		"TRELLO_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TRELLO_TEST_ACTION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TRELLO_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TRELLO_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTrelloSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRELLO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRELLO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
